espnet2/asr_transducer/encoder/modules/positional_encoding.py

```diff
@@ -0,0 +1,91 @@
+"""Positional encoding modules."""
+
+import math
+
+import torch
+
+from espnet.nets.pytorch_backend.transformer.embedding import _pre_hook
+
+
+class RelPositionalEncoding(torch.nn.Module):
+    """Relative positional encoding.
+
+    Args:
+        size: Module size.
+        max_len: Maximum input length.
+        dropout_rate: Dropout rate.
+
+    """
+
+    def __init__(
+        self, size: int, dropout_rate: float = 0.0, max_len: int = 5000
+    ) -> None:
+        """Construct a RelativePositionalEncoding object."""
+        super().__init__()
+
+        self.size = size
+
+        self.pe = None
+        self.dropout = torch.nn.Dropout(p=dropout_rate)
+
+        self.extend_pe(torch.tensor(0.0).expand(1, max_len))
+        self._register_load_state_dict_pre_hook(_pre_hook)
+
+    def extend_pe(self, x: torch.Tensor, left_context: int = 0) -> None:
+        """Reset positional encoding.
+
+        Args:
+            x: Input sequences. (B, T, ?)
+            left_context: Number of frames in left context.
+
+        """
+        time1 = x.size(1) + left_context
+
+        if self.pe is not None:
+            if self.pe.size(1) >= time1 * 2 - 1:
+                if self.pe.dtype != x.dtype or self.pe.device != x.device:
+                    self.pe = self.pe.to(device=x.device, dtype=x.dtype)
```

Comment:
I'm not sure if there are cases where `self.pe` and inputs have different `device` or `dytpe` properties?

Reply:
Yes, I'm dumb.

Reply:
I think we can remove this if statement and the `to()` statement inside it.

Reply:
I thought so too but I had a mismatch at runtime. I need to do a complete trace because there are unnecessary cast or device pining along the way.